precompiles/precompile-registry/src/lib.rs

```diff
@@ -0,0 +1,71 @@
+// Copyright 2019-2022 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+use core::marker::PhantomData;
+use fp_evm::PrecompileSet;
+use precompile_utils::{precompile_set::IsActivePrecompile, prelude::*};
+use sp_core::Get;
+
+const DUMMY_CODE: [u8; 5] = [0x60, 0x00, 0x60, 0x00, 0xfd];
+
+pub struct PrecompileRegistry<Runtime>(PhantomData<Runtime>);
+
+#[precompile_utils::precompile]
+impl<Runtime> PrecompileRegistry<Runtime>
+where
+	Runtime: pallet_evm::Config,
+	Runtime::PrecompilesType: IsActivePrecompile,
+{
+	#[precompile::public("isPrecompile(address)")]
+	#[precompile::view]
+	fn is_precompile(_handle: &mut impl PrecompileHandle, address: Address) -> EvmResult<bool> {
+		// TODO: what cost?
```

Comment:
I think we made a mistake when we designed the signature of `is_precompile` from the beginning, we should refactor `is_precompile` to add the parameter `handle: &mut impl PrecompileHandle` and return `EvmResult<bool>` instead of `bool`, this is the proper way to account gas consumption.
We also need top refactor `discrimimant` and `is_evm_proxy_call_allowed` the same way.

Reply:
for `is_precompile` is the `PrecompileSet` trait it cannot be changed, as it comes from the EVM and is called to know if an address is cold or warm (precompile addresses are always warm). Nothing can be charged for that. However we can consider it is covered by the gas cost of doing a subcall.

In this precompile it is an additional call to this function so it should be charged again. What we can do is add a new trait/function for that in the util, taking an Option of the precompile handle, and have the EVM `is_precompile` call it with None while from this precompile we call it with the handle.

Reply:
> However we can consider it is covered by the gas cost of doing a subcall.

I don't think so

> What we can do is add a new trait/function for that in the util, taking an Option of the precompile handle, and have the EVM is_precompile call it with None while from this precompile we call it with the handle

Definitly a good idea

Reply:
Thing is that precompiles were not originally made to be as dynamic as in Moonbeam. We have multiple checks to reduce the cost of knowing if a given address is one of a precompile, but for exemple if you try to call a smart contract whose address starts with 0xffffffff it will require the AssetsErc20 precompile set to do a storage read to see if the address corresponds to an XC20 token first. Thus calling this contract which induce an additional cost just because it starts with this prefix?

A possible solution to consider this check cost negligible is to read the account code first and return false immediately if it is neither empty or dummy code. This would be negligible because afterward the EVM will fetch the code too (which is already cached by Substrate). However there would still be uncharged cost if one call an address that is in the range but is neither a SC nor an existing asset precompile.

We can make a PR to the EVM repo to add an `&mut impl PrecompileHandle` to charge gas in that last case if you think that is necessary.

Reply:
We must make sure that all costs are taken into account (even if they are overestimated in some cases), otherwise an attacker can spam by paying less than he should.

The safest solution is to be too restrictive, only match precompiles by range and prefix, so if a contract starts with `0xffffffff` it will not be supported.

The other solution is to make the necessary db reads, but this implies increasing the base cost of each evm call to take into account the potential cost of these checks.

There is no good solution unfortunately, I think we made a mistake from the beginning by allowing "dynamic" precompiles with an prefix of only 4 bytes.

Reply:
> I think we made a mistake from the beginning by allowing "dynamic" precompiles with an prefix of only 4 bytes.
I would say it is the opposite, as without the prefix it would require 2 storage read each time `is_precompile` is called, while currently it is called only if the address is in that range. Another approach would have been to have a single precompile, however then it wouldn't have complied with the ERC20 standard (it could have complied with ERC1155 instead).

Anyway I can make a PR in the EVM repo to allow charging for `is_precompile`.

Reply:
https://github.com/rust-blockchain/evm/pull/157

Reply:
I discussed with Alan and recording the cost of is_precompile in the EVM might break some Ethereum tooling and our tests, as doing a transfer to an account will no longer always cost 21000 gas.

To move forward I now record 1 db read inside the registry. We'll go back to this once we've decided what to do with is_precompile cost and the change is reflected to the EVM repo and frontier.

Reply:
> recording the cost of is_precompile in the EVM might break some Ethereum tooling and our tests, as doing a transfer to an account will no longer always cost 21000 gas.

Yes, doing a transfer to an account that start with `0xFFFFFFFF..` will cost more than 21000, but for accounts without this prefix the cost should still 21000.

Reply:
To move forward we'll only record the 1 db read cost inside this precompile, and we'll change it in a later PR while we figure out the proper thing to do with is_precompile cost inside the EVM (current idea of recording the cost breaks the assumption that all transfers cost 21000)

Reply:
> breaks the assumption that all transfers cost 21000

technically we can make sure that this assumption remains guaranteed even counting the real cost of `is_precompile`. Technically we can make sure that this assumption remains guaranteed even if we count the cost of is_precompile.

We have to modify the implementation of the evm to handle this case, but let's leave it for a future ticket.